gobblin-core-base/src/main/java/gobblin/crypto/RotatingAESCodec.java

```diff
@@ -0,0 +1,221 @@
+
+
+/**
+ * Implementation of an encryption algorithm that works in the following way:
+ *
+ * 1. A credentialStore is provisioned with a set of AES keys
+ * 2. When wrapOutputStream() is called, an AES key will be picked at random and a new initialization vector (IV)
+ *    will be generated.
+ * 3. A header will be written [keyId][ivLength][base64 encoded iv]
+ * 4. Ciphertext will be base64 encoded and written out. We do not insert linebreaks.
+ */
+@Slf4j
+public class RotatingAESCodec implements StreamCodec {
+  private static final int AES_KEY_LEN = 16;
+
+  private final Random random;
+  private final CredentialStore credentialStore;
+  private List<KeyRecord> keyRecords;
+
+  /**
+   * Create a new encryptor
+   * @param credentialStore Credential store where keys can be found
+   */
+  public RotatingAESCodec(CredentialStore credentialStore) {
+    this.credentialStore = credentialStore;
+    this.random = new Random();
+  }
+
+  @Override
+  public OutputStream wrapOutputStream(OutputStream origStream) {
+    return new StreamInstance(random, getKeyRecords(), origStream).wrapOutputStream();
+  }
+
+  @Override
+  public InputStream wrapInputStream(InputStream origStream) throws IOException {
+    throw new AssertionError("not implemented yet");
+  }
+
+  private synchronized List<KeyRecord> getKeyRecords() {
+    if (keyRecords == null) {
+      keyRecords = new ArrayList<>();
+      for (Map.Entry<String, byte[]> entry : credentialStore.getAllEncodedKeys().entrySet()) {
+        if (entry.getValue().length != AES_KEY_LEN) {
+          log.debug("Skipping keyId {} because it is length {}; expected {}",
+              entry.getKey(),
+              entry.getValue().length,
+              AES_KEY_LEN);
+          continue;
+        }
+
+        try {
+          Integer keyId = Integer.parseInt(entry.getKey());
+          SecretKey key = new SecretKeySpec(entry.getValue(), "AES");
+
+          keyRecords.add(new KeyRecord(keyId, key));
+        } catch (NumberFormatException e) {
+          log.debug("Skipping keyId {} because this algorithm can only use numeric key ids", entry.getKey());
+        }
+      }
+    }
+
+    return keyRecords;
+  }
+
+  @Override
+  public String getTag() {
+    return "aes_rotating";
+  }
+
+  /**
+   * Represents a set of parsed AES keys that we can choose from when encrypting.
+   */
+  static class KeyRecord {
+    private final int keyId;
+    private final SecretKey secretKey;
+
+    KeyRecord(int keyId, SecretKey secretKey) {
+      this.keyId = keyId;
+      this.secretKey = secretKey;
+    }
+
+    int getKeyId() {
+      return keyId;
+    }
+
+    SecretKey getSecretKey() {
+      return secretKey;
+    }
+  }
+
+  /**
+   * Helper class that keeps state around for a wrapped output stream. Each stream will have a different
+   * selected key, IV, and cipher state.
+   */
+  static class StreamInstance {
+    private final OutputStream origStream;
+    private final List<KeyRecord> keyRecords;
```

Comment:
Why is `keyRecords` and `random` needed here? `initCipher` could be done by the parent class.